### PR TITLE
Add allowedChildren call to Outgoing Editor events

### PR DESCRIPTION
**Pull request by patrickdemooij9**
This is a ported version of this v8 PR: https://github.com/umbraco/Umbraco-CMS/pull/9906
Most of the description of that PR still applies to this one. Although, the code example is a bit different now.

https://user-images.githubusercontent.com/11466511/109339815-77656e00-7868-11eb-8aef-3a93076da116.gif
Can be done with the following handler:
```
public class TestCode : INotificationHandler<SendingAllowedChildrenNotification>
    {
        public void Handle(SendingAllowedChildrenNotification notification)
        {
            var children = notification.Children.ToArray();
            if (children.All(it => it.Alias != "settings"))
                return;

            var currentSettings = notification.UmbracoContext.Content.GetSingleByXPath("//settings");
            if (currentSettings is null)
                return;

            notification.Children = children.Where(it => it.Alias != "settings");
        }
    }
```

I think this feature is a long requested feature and would also allow package developers to create UI's based on this.

### CODE REVIEW

**Comment by mikecp** [comment]
Hello @patrickdemooij9 , thanks for this PR, it adds a great functionality 👍

I reviewed it quickly to avoid it getting closed again when we go to V10 or something 😅😜

I want to do a last test to see what happens if the notification handler actually changes the collection of allowed children e.g. by adding a doc type that is actually not allowed, but up to now all looks good!

**Comment by mikecp** [comment]
Hello @patrickdemooij9 ,

I hope you had a nice week-end 😉
Keeping you up to date: I have tested and it is indeed possible to use the notification handler to add not allowed doc types in the children lists; and then to actually create pages of that type as children, because apparently there is no check at Save time (probably because this was not possible yet 😁)
I am waiting for feedback from HQ about how this should be handled (or not), I'll keep you posted.

File: src/Umbraco.Web.BackOffice/Filters/OutgoingEditorModelEventAttribute.cs
Comment (patrickdemooij9) [reply src/Umbraco.Web.BackOffice/Filters/OutgoingEditorModelEventAttribute.cs:88]: Yeah, I wasn't sure if that was something we would want as it currently has no setter, only a getter. If we want it to be the same, I'll gladly change it!

**Review by mikecp — APPROVED** [review]
Hey @patrickdemooij9 , I got the green light from HQ, so let's merge this change 👍🎉

File: src/Umbraco.Web.BackOffice/Filters/OutgoingEditorModelEventAttribute.cs
Comment (mikecp) [reply src/Umbraco.Web.BackOffice/Filters/OutgoingEditorModelEventAttribute.cs:88]: Hey @ronaldbarendse @patrickdemooij9 , it might be an interesting idea indeed, but if we go for it I would  suggest implementing it in another PR 😉

**Comment by mikecp** [comment]
FYI: Errors are  not related to the changes in here.

**Comment by bjarnef** [comment]
@patrickdemooij9 it would be great to add an example of this to the documentation. I couldn't find anything documented regarding `SendingAllowedChildrenNotification`, but we have this for v9+
https://our.umbraco.com/documentation/reference/Notifications/EditorModel-Notifications/

If we want to restrict/limit multiple document types I wonder if we would have a notification handler per document type or if could be combined in same notification handler, e.g. using a dictionary with alias as key and amount as value.

**Pull request merged**
